lgc/state/PalMetadata.cpp

```diff
@@ -872,3 +867,95 @@ void PalMetadata::setFinalized128BitCacheHash(const Hash128 &finalizedCacheHash,
+// Currently this just copies the fragment shader input mapping information.
+// Storing it in our PAL metadata, even though it will be erased before the PAL metadata is written into the ELF,
+// means that it persists in the IR if compilation is interrupted by e.g. -emit-lgc.
+//
+// @param other : PAL metadata object for "other" pipeline
+void PalMetadata::setOtherPartPipeline(PalMetadata *other) {
+  m_pipelineNode[PipelineMetadataKey::FragInputMapping1] =
+      other->m_pipelineNode[PipelineMetadataKey::FragInputMapping1];
+  m_pipelineNode[PipelineMetadataKey::FragInputMapping2] =
+      other->m_pipelineNode[PipelineMetadataKey::FragInputMapping2];
+}
+
+// =====================================================================================================================
+// Check whether we have FS input mappings, and thus whether we're doing part-pipeline compilation of the
+// pre-FS part of the pipeline.
+bool PalMetadata::haveFsInputMappings() {
+  return m_pipelineNode.find(m_document->getNode(PipelineMetadataKey::FragInputMapping1)) != m_pipelineNode.end();
+}
+
+// =====================================================================================================================
+// In part-pipeline compilation, get a blob of data representing the FS input mappings that can be used by the
+// client in a hash. The resulting data is owned by the PalMetadata object, and lives until the PalMetadata
+// object is destroyed or another call is made to getFsInputMappings.
+StringRef PalMetadata::getFsInputMappings() {
+  assert(haveFsInputMappings());
+  m_fsInputMappingsBlob.clear();
+  auto fragInputMappingArray1 = m_pipelineNode[PipelineMetadataKey::FragInputMapping1].getArray(true);
+  auto fragInputMappingArray2 = m_pipelineNode[PipelineMetadataKey::FragInputMapping2].getArray(true);
+  for (auto &element : fragInputMappingArray1) {
+    unsigned elementVal = element.getUInt();
+    m_fsInputMappingsBlob.append(StringRef(reinterpret_cast<const char *>(&elementVal), sizeof(elementVal)));
+  }
+  for (auto &element : fragInputMappingArray2) {
+    unsigned elementVal = element.getUInt();
+    m_fsInputMappingsBlob.append(StringRef(reinterpret_cast<const char *>(&elementVal), sizeof(elementVal)));
+  }
+  return m_fsInputMappingsBlob;
+}
+
+// =====================================================================================================================
+// In part-pipeline compilation, retrieve the FS input mappings into the provided vectors.
+//
+// This function copes with that metadata not being present, as it can be called when doing a compile of a single
+// shader in the command-line tool.
+//
+// @param [out] data1 : Vector for first array of values
+// @param [out] data2 : Vector for second array of values
+void PalMetadata::retrieveFragmentInputInfo(SmallVectorImpl<unsigned> &data1, SmallVectorImpl<unsigned> &data2) {
```

Comment:
Again can we get more meaningful names and descriptions in the comments.  Somebody looking at this will have not idea what to do with data1 and data2 or how to interpret it.